File: 04-how-sum/01.py
# In this problem, we take a target, and return a combination of numbers array.
# Return the operands which sum to target.


def howSumHelper(lst, target, stk):
    if target == 0:
        return stk

    if target < 0:
        return None

    for num in lst:
        rem = target - num

        stk.append(num)
        if howSumHelper(lst, rem, stk):
            return stk

        stk.pop()

    return None


def howSum(lst, target):
    return howSumHelper(lst, target, [])


if __name__ == '__main__':
    print(howSum([2, 3], 7))
    print(howSum([5, 3, 4, 7], 7))
    print(howSum([2, 4], 7))
    print(howSum([2, 3, 5], 8))
    print(howSum([7, 14], 300))

    # True, True, False, True, False

File: 04-how-sum/01.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> howSumHelper(vector<int> &lst, int target, vector<int> stk)
{
    if (target < 0)
        return vector<int>{};

    if (target == 0)
        return stk;

    for (auto num : lst)
    {
        stk.push_back(num);
        vector<int> ans = howSumHelper(lst, target - num, stk);
        if (ans.size() != 0)
            return ans;

        stk.pop_back();
    }

    return vector<int>{};
}

void howSum(vector<int> lst, int target)
{
    vector<int> ret = howSumHelper(lst, target, vector<int>{});

    if (ret.size() == 0)
    {
        cout << "No Possible Combination" << endl;
        return;
    }

    cout << "One Possible Combination is:" << endl;
    for (auto num : ret)
        cout << num << " ";
    cout << endl;
}

int main()
{
    howSum(vector<int>{2, 3}, 7);
    howSum(vector<int>{5, 3, 4, 7}, 7);
    howSum(vector<int>{2, 4}, 7);
    howSum(vector<int>{2, 3, 5}, 8);
    howSum(vector<int>{7, 14}, 300);

    return 0;
}